packages/engine/src/datastore/batch/mod.rs

```diff
@@ -9,105 +9,405 @@ pub mod message;
+    pub fn flush_changes(&mut self) -> Result<()> {
+        if !self.is_persisted() {
+            return Err(Error::from(format!(
+                "Tried to flush changes older than or equal to already written data: {:?}, {:?}",
+                self.loaded_metaversion(),
+                self.persisted_metaversion(),
+            )));
+        }
+
+        let changes = std::mem::take(&mut self.changes);
+        let changed = GrowableBatch::flush_changes(self, changes)?;
 
+        let mut persisted = self.persisted_metaversion();
+        let before = persisted;
+        if changed.resized() {
+            persisted.increment(); // Memory must be reloaded (along with batch).
+        } else {
+            persisted.increment_batch(); // Only batch needs to be reloaded.
+        };
+        // TODO: Loaded memory version is always up to date here,
+        //       because if shared memory was resized in flush_changes,
+        //       that automatically means it was mapped again in this
+        //       process.
+        self.segment.set_persisted_metaversion(persisted);
+        tracing::debug!(
+            "Flush metaversions: {before:?}, {persisted:?}, {:?}",
+            self.persisted_metaversion()
+        );
         Ok(())
     }
 
-    fn sync(&mut self, batch: &Self) -> Result<()> {
-        if self.memory().size < batch.memory().size {
-            self.memory_mut().resize(batch.memory().size)?;
-            self.metaversion_mut().increment();
+    /// Add an Arrow column's data to an internal queue of changes
+    /// to later persist to memory.
+    ///
+    /// # Errors
+    ///
+    /// If the loaded metaversion isn't equal to the
+    /// persisted metaversion, this gives an error to avoid
+    /// queueing stale data. (The loaded metaversion can't
+    /// be newer than the persisted metaversion.)
+    /// (TODO: We might have to remove this restriction to
+    ///        allow flushing changes multiple times without
+    ///        loading.)
+    pub fn queue_change(&mut self, change: ColumnChange) -> Result<()> {
+        if self.is_persisted() {
+            self.queue_change_unchecked(change)
+        } else {
+            Err(Error::from(format!(
+                "Tried to queue changes older than or equal to already written data: {:?}, {:?}",
+                self.loaded_metaversion(),
+                self.persisted_metaversion(),
+            )))
+        }
+    }
+
+    /// Return record batch if and only if the loaded version
+    /// is equal to the persisted version. (The loaded version
+    /// can't be newer than the persisted version.)
+    pub fn record_batch(&self) -> Result<&RecordBatch> {
+        if self.is_persisted() {
+            Ok(self.record_batch_unchecked())
+        } else {
+            assert!(false, "ALoaded record batch is older than persisted one");
+            Err(Error::from(
+                "Loaded record batch is older than persisted one",
+            ))
+        }
```

Comment:
```suggestion
        assert(self.is_persisted(), "Loaded record batch is older than persisted one");
        self.record_batch_unchecked()
```

Reply:
Sorry, this was left over from debugging